linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala

```diff
@@ -1518,19 +1518,68 @@ private class FunctionEmitter private (
     val BinaryOp(op, lhs, rhs) = tree
     assert(op == === || op == !==)
 
-    // TODO Optimize this when the operands have a better type than `any`
+    def maybeGenInvert(): Unit = {
+      if (op == BinaryOp.!==)
+        fb += wa.I32Eqz
+    }
 
-    genTree(lhs, AnyType)
-    genTree(rhs, AnyType)
+    val lhsType = lhs.tpe
+    val rhsType = rhs.tpe
 
-    markPosition(tree)
+    if (lhsType == NothingType) {
+      genTree(lhs, NothingType)
+      NothingType
+    } else if (rhsType == NothingType) {
+      genTree(lhs, NoType)
+      genTree(rhs, NothingType)
+      NothingType
+    } else if (rhsType == NullType) {
+      /* Note that the optimizer normalizes Literals on the right of `===`,
+       * so testing for the `lhsType == NullType` is not as useful.
+       */
+      genTree(lhs, AnyType)
+      genTree(rhs, NoType) // no-op if it is actually a Null() literal
+      markPosition(tree)
+      fb += wa.RefIsNull
+      maybeGenInvert()
+      BooleanType
+    } else {
+      val lhsWasmType = transformSingleType(lhsType)
+      val rhsWasmType = transformSingleType(rhsType)
 
-    fb += wa.Call(genFunctionID.is)
+      /* Is the given heap type a sub-heap-type of `eq`, assuming that it was
+       * taken from the result of `transformSingleType`?
+       *
+       * Note that all the `HeapType.Type(_)`s returned by
+       * `transformSingleType` point to struct types, which are sub-heap-types
+       * of `eq`.
+       *
+       * In practice, the only case that is *not* a sub-heap-type of `eq` is
+       * `any`.
+       */
```

Comment:
This helps, TY! Could you also add a sentence explaining that `Object.is` coincides with `ref.eq` for all `eq` types in our encoding? IMO this is non-obvious. (or in other words: this explains why `ref.eq` can be used, but it doesn't explain why it is correct to be used).

From a code-flow POV, I would even go farther with helpers. Have you considered something like this:

```scala
/* Is the given type a sub-heap-type of `eq`, ...  */
def canRefEq(tpe: Type): Boolean = transformSingleType(tpe) match {
  case watpe.RefType(_, heapType) => heapType != watpe.HeapType.Any
  case  _                         => false
}

if (lhsType == NothingType) {
  // snip
} else if (rhsType == /* snip */) {
  // snip
} else if (canRefEq(lhsType) && canRefEq(rhsType)) {
  // TODO: Explanation why RefEq is correct.
  genTree(lhs, lhsType)
  genTree(rhs, rhsType)
  markPosition(tree)
  fb += wa.RefEq
} else {
  // Fall back on the `Object.is` helper
  genTree(lhs, AnyType)
  genTree(rhs, AnyType)
  markPosition(tree)
  fb += wa.Call(genFunctionID.is)
}
```

Actually, have you considered putting the `isEq` check into `webassembly.Types`? It is clearly a WASM core thing
and the fact that `heapType != Any` ==> `heapType <: eq` seems to be a consequence of our encoding of WASM types.

Reply:
> Actually, have you considered putting the `isEq` check into `webassembly.Types`? It is clearly a WASM core thing
> and the fact that `heapType != Any` ==> `heapType <: eq` seems to be a consequence of our encoding of WASM types.

That would not be correct. In general there are other heap types that are not subtypes of `eq`. For example `func`, as well as all `TypeID`s that point to `func` types. Hence the part of the comment that says

> Note that all the `HeapType.Type(_)`s returned by `transformSingleType` point to struct types, which are sub-heap-types of `eq`.

Reply:
> This helps, TY! Could you also add a sentence explaining that `Object.is` coincides with `ref.eq` for all `eq` types in our encoding? IMO this is non-obvious. (or in other words: this explains why `ref.eq` can be used, but it doesn't explain why it is correct to be used).
> 
> From a code-flow POV, I would even go farther with helpers. Have you considered something like this:

Done and done.